protocols/mempool-overview.md

```diff
@@ -134,10 +139,134 @@ Committed transactions are not in the mempool:
+
+Another practical concern is with INV3.
+This invariant requires to have a green light from the client application before adding a transaction to the mempool.
+For efficiency, such a validation needs to be made at most $\beta$ times per transaction at each height, for some parameter $\beta$.
+Ideally, $\beta$ equals $1$.
+In practice, $\beta = f(T)$ for some function f of the maximal number of transactions T submitted between two heights.
+Given some transaction $tx$, variable $p.valid[tx]$ tracks the number of times the application was asked at the current height.
+Invariant INV3 is replaced with:  
+**INV3a.** $\forall tx. \forall p \in Correct. \square(tx \in p.mempool \implies p.valid[tx] \in [1, \beta])$
+
+
+## Implementation in CometBFT (as of v0.38.0-alpha.2)
+
+The mempool is implemented in `clist_mempool.go`, in the `CListMempool` data type.
+`CListMempool` uses a single variable for the two mechanisms covered in the previous section.
+Below, we present this approach in detail then establish its correctness.
+
+**Algorithm**
+For starters, we explain at a high-level the logic in `CListMempool`.
+Variables `txs` and `cache` respectively hold the mempool and the validity cache in a FIFO and LRU list.
+We omit the use of `txsMap`, assuming that a transaction is in `txs` iff it is also in `txsMap` (see [this](https://github.com/cometbft/cometbft/pull/890) fix).
+For simplicity, we consider that 
+_(i)_ the mempool is never full, 
+_(ii)_ invalid transactions are _not_ kept in the cache, that is the parameter `KeepInvalidTxsInCache` is always set to false, and
+_(iii)_ the `flush` operation is never called.
+In addition, we shall assume that a finite amount of transactions is received by the system, and that if a transaction is received at a correct process, then it is eventually received at all the correct processes (thanks to the gossip layer in `mempool/reactor.go`).
+
+Then, let us consider some transaction $tx$.
+According to the logic in `clist_mempool.go`,
+
+$tx$ is added to the cache at time $t$ (in short, @t) if  
+(1) $tx$ is received and not already there @t (`checkTx`, [l237](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L237)), or  
+(2) $tx$ is committed and valid @$t (`update`, [l598](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L598)).
+
+$tx$ is removed from the cache @t if  
+(3) $tx$ was received @t'<t, re-checked and invalid t'<@t''<t (`resCbRecheck`, [l477](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L4747)), or  
+(4) $tx$ is ejected from the cache as it is the last transaction wrt. cache and a new transaction is added, or  
+(5) $tx$ is committed and invalid @t'<t (`update`, [l598](https://github.com/cometbft/cometbft/blob/1f524d12996204f8fd9d41aa5aca215f80f06f5e/mempool/clist_mempool.go#L601)).
```

Comment:
"and invalid". I remember having a discussion with @hvanz on this. My impression at the time was that we probably don't want that.

Do you have a clearer idea now on whether we want to remove failed, committed TXs from the cache?